// type-dom.ts/type-element/type-html/index.js
export { TypeA } from './a/a.abstract';
export { TypeAbbr } from './abbr/abbr.abstract';
export { TypeAddress } from './address/address.abstract';
export { TypeArea } from './area/area.abstract';
export { TypeArticle } from './article/article.abstract';
export { TypeAside } from './aside/aside.abstract';
export { TypeAudio } from './audio/audio.abstract';
export { TypeB } from './b/b.abstract';
export { TypeBase } from './base/base.abstract';
export { TypeBdi } from './bdi/bdi.abstract';
export { TypeBdo } from './bdo/bdo.abstract';
export { TypeBlockQuote } from './block-quote/block-quote.abstract';
export { TypeBr } from './br/br.abstract';
export { TypeButton } from './button/button.abstract';
export { TypeCanvas } from './canvas/canvas.abstract';
export { TypeCite } from './cite/cite.abstract';
export { TypeCode } from './code/code.abstract';
export { TypeData } from './data/data.abstract';
export { TypeDataList } from './data-list/data-list.abstract';
export { TypeDel } from './del/del.abstract';
export { TypeDetails } from './details/details.abstract';
export { TypeDfn } from './dfn/dfn.abstract';
export { TypeDialog } from './dialog/dialog.abstract';
export { TypeDiv } from './div/div.abstract';
export { TypeDL } from './dl/dl.abstract';
export { TypeDD } from './dl/dd/dd.abstract';
export { TypeDT } from './dl/dt/dt.abstract';
export { TypeEm } from './em/em.abstract';
export { TypeEmbed } from './embed/embed.abstract';
export { TypeFieldset } from './fieldset/fieldset.abstract';
export { TypeFigCaption } from './fig-caption/fig-caption.abstract';
export { TypeFigure } from './figure/figure.abstract';
export { TypeFooter } from './footer/footer.abstract';
export { TypeForm } from './form/form.abstract';
export { TypeHead } from './head/head.abstract';
export { TypeHeader } from './header/header.abstract';
export { TypeHGroup } from './hgroup/hgroup.abstract';
export { TypeHr } from './hr/hr.abstract';
export { TypeI } from './i/i.abstract';
export { TypeIFrame } from './iframe/iframe.abstract';
export { TypeImg } from './img/img.abstract';
export { TypeInput } from './input/input.abstract';
export { TypeIns } from './ins/ins.abstract';
export { TypeKbd } from './kbd/kbd.abstract';
export { TypeLabel } from './label/label.abstract';
export { TypeLegend } from './legend/legend.abstract';
export { TypeMain } from './main/main.abstract';
export { TypeMap } from './map/map.abstract';
export { TypeMark } from './mark/mark.abstract';
export { TypeMenu } from './menu/menu.abstract';
export { TypeMenuItem } from './menu-item/menu-item.abstract';
export { TypeMeter } from './meter/meter.abstract';
export { TypeNav } from './nav/nav.abstract';
export { TypeOL } from './ol/ol.abstract';
export { TypeOptGroup } from './opt-group/opt-group.abstract';
export { TypeOption } from './option/option.abstract';
export { TypeOutput } from './output/output.abstract';
export { TypeP } from './p/p.abstract';
export { TypePicture } from './picture/picture.abstract';
export { TypePre } from './pre/pre.abstract';
export { TypeProgress } from './progress/progress.abstract';
export { TypeQ } from './q/q.abstract';
export { TypeRp } from './rp/rp.abstract';
export { TypeRt } from './rt/rt.abstract';
export { TypeRuby } from './ruby/ruby.abstract';
export { TypeS } from './s/s.abstract';
export { TypeSamp } from './samp/samp.abstract';
export { TypeSection } from './section/section.abstract';
export { TypeSelect } from './select/select.abstract';
export { TypeSmall } from './small/small.abstract';
export { TypeSource } from './source/source.abstract';
export { TypeSpan } from './span/span.abstract';
export { TypeStrong } from './strong/strong.abstract';
export { TypeSub } from './sub/sub.abstract';
export * from './table';
export { TypeTextarea } from './textarea/textarea.abstract';
export { TypeTime } from './time/time.abstract';
export { TypeTitle } from './title/title.abstract';
export { TypeTrack } from './track/track.abstract';
export { TypeU } from './u/u.abstract';
export { TypeUL } from './ul/ul.abstract';
export { TypeLI } from './ul/li/li.abstract';
export { TypeVar } from './var/var.abstract';
export { TypeVideo } from './video/video.abstract';
export { TypeWbr } from './wbr/wbr.abstract';
