// type-dom.ts/style/style.enum.js
export var TextDecoration;
(function (TextDecoration) {
    TextDecoration["none"] = "none";
    TextDecoration["underline"] = "underline";
    TextDecoration["overline"] = "overline";
    TextDecoration["lineThrough"] = "line-through";
    TextDecoration["blink"] = "blink";
    TextDecoration["inherit"] = "inherit";
})(TextDecoration || (TextDecoration = {}));
export var JustifyContent;
(function (JustifyContent) {
    JustifyContent["left"] = "left";
    JustifyContent["center"] = "center";
    JustifyContent["right"] = "right";
    JustifyContent["both"] = "both";
    JustifyContent["mediumKashida"] = "medium-kashida";
    JustifyContent["distribute"] = "distribute";
    JustifyContent["numTab"] = "num-tab";
    JustifyContent["highKashida"] = "high-kashida";
    JustifyContent["lowKashida"] = "low-kashida";
    JustifyContent["thaiDistribute"] = "thai-distribute";
    JustifyContent["horizontal"] = "horizontal";
})(JustifyContent || (JustifyContent = {}));
export var LineSpacingRule;
(function (LineSpacingRule) {
    LineSpacingRule["auto"] = "auto";
    LineSpacingRule["exact"] = "exact";
    LineSpacingRule["atLeast"] = "atLeast";
})(LineSpacingRule || (LineSpacingRule = {}));
export var OnOff;
(function (OnOff) {
    OnOff["true"] = "true";
    OnOff["false"] = "false";
    OnOff["on"] = "on";
    OnOff["off"] = "off";
})(OnOff || (OnOff = {}));
export var FontTheme;
(function (FontTheme) {
    FontTheme["majorEastAsia"] = "major-east-asia";
    FontTheme["majorBidi"] = "major-bidi";
    FontTheme["majorAscii"] = "major-ascii";
    FontTheme["majorHAnsi"] = "majorHAnsi";
})(FontTheme || (FontTheme = {}));
export var StylePosition;
(function (StylePosition) {
    StylePosition["absolute"] = "absolute";
    StylePosition["fixed"] = "fixed";
    StylePosition["relative"] = "relative";
    StylePosition["static"] = "static";
    StylePosition["inherit"] = "inherit";
})(StylePosition || (StylePosition = {}));
export var Display;
(function (Display) {
    Display["none"] = "none";
    Display["block"] = "block";
    Display["inline"] = "inline";
    Display["inlineBlock"] = "inline-block";
    Display["listItem"] = "list-item";
    Display["runIn"] = "run-in";
    Display["compact"] = "compact";
    Display["marker"] = "marker";
    Display["table"] = "table";
    Display["inlineTable"] = "inline-table";
    Display["tableRowGroup"] = "table-row-group";
    Display["tableHeaderGroup"] = "table-header-group";
    Display["tableFooterGroup"] = "table-footer-group";
    Display["tableRow"] = "table-row";
    Display["tableColumnGroup"] = "table-column-group";
    Display["tableColumn"] = "table-column";
    Display["tableCell"] = "table-cell";
    Display["tableCaption"] = "table-caption";
    Display["inherit"] = "inherit";
    Display["flex"] = "flex";
})(Display || (Display = {}));
export var Cursor;
(function (Cursor) {
    Cursor["default"] = "default";
    Cursor["auto"] = "auto";
    Cursor["crosshair"] = "crosshair";
    Cursor["pointer"] = "pointer";
    Cursor["move"] = "move";
    Cursor["eResize"] = "e-resize";
    Cursor["neResize"] = "ne-resize";
    Cursor["nwResize"] = "nw-resize";
    Cursor["nResize"] = "n-resize";
    Cursor["seResize"] = "se-resize";
    Cursor["swResize"] = "sw-resize";
    Cursor["sResize"] = "s-resize";
    Cursor["wResize"] = "w-resize";
    Cursor["text"] = "text";
    Cursor["wait"] = "wait";
    Cursor["help"] = "help";
    Cursor["notAllowed"] = "not-allowed";
})(Cursor || (Cursor = {}));
